### PR TITLE
glib: fix tests

### PR DESCRIPTION
**Pull request by alyssais**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by alyssais** [comment]
@OfBorg build glib.override{doCheck=true;}

**Comment by risicle** [comment]
Could use the `finalAttrs.finalPackage` trick to add a `passthru.tests.withChecks` which would allow you to call it more easily (and perhaps make it more likely for people to try it)

**Comment by alyssais** [comment]
> Could use the `finalAttrs.finalPackage` trick to add a `passthru.tests.withChecks` which would allow you to call it more easily (and perhaps make it more likely for people to try it)

How would I do that?  I tried

```
    tests.withChecks = finalAttrs.finalPackage.override { doCheck = true; };
```

But this doesn't work because `finalPackage` is not overridable.

File: pkgs/development/libraries/glib/default.nix [hunk 1]
```diff
@@ -154,11 +161,25 @@ stdenv.mkDerivation rec {
     patchShebangs glib/gen-unicode-tables.pl
     patchShebangs glib/tests/gen-casefold-txt.py
     patchShebangs glib/tests/gen-casemap-txt.py
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/glib/default.nix:163]:
```suggestion
    patchShebangs glib/gen-unicode-tables.pl \
      glib/tests/gen-casefold-txt.py \
      glib/tests/gen-casemap-txt.py
```
Unrelated but we can simplify this

File: pkgs/development/libraries/glib/default.nix [hunk 2]
```diff
@@ -154,11 +161,25 @@ stdenv.mkDerivation rec {
...
+    # Needed because of libtool wrappers
+    sed -e '/g_subprocess_launcher_set_environ (launcher, envp);/a g_subprocess_launcher_setenv (launcher, "PATH", g_getenv("PATH"), TRUE);' -i gio/tests/gsubprocess.c
   '' + lib.optionalString stdenv.hostPlatform.isWindows ''
     substituteInPlace gio/win32/meson.build \
       --replace "libintl, " ""
   '';
 
+  postConfigure = ''
+    patchShebangs gio/gdbus-2.0/codegen/gdbus-codegen gobject/glib-{genmarshal,mkenums}
+  '';
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/glib/default.nix:181]: This should be done in postPatch

**Comment by risicle** [comment]
> But this doesn't work because `finalPackage` is not overridable.

Oh yes, can't use `.override`, but you should be able to use `.overrideAttrs`.

File: pkgs/development/libraries/glib/default.nix (hunk 2 above)
Comment (alyssais) [reply pkgs/development/libraries/glib/default.nix:181]: Please read the commit messages.

File: pkgs/development/libraries/glib/default.nix (hunk 1 above)
Comment (alyssais) [reply pkgs/development/libraries/glib/default.nix:163]: This PR is big enough as it is, so I don't want to start adding in unrelated changes.

**Review by 7c6f434c — APPROVED** [review]
Diff looks good

**Comment by SuperSandro2000** [comment]
Some tests seem to only expected fail on x86_64-linux, not any other platform. Maybe ignore those on other platforms?

**Comment by alyssais** [comment]
> Some tests seem to only expected fail on x86_64-linux, not any other platform. Maybe ignore those on other platforms?

I'm not sure what you mean, but my goal here is just to make it possible to run the tests at all, not to make them all pass.

**Review by risicle — APPROVED** [review]

**Comment by jtojnar** [comment]
Changes look good overall but it might be better to check for `ST_NOATIME` as suggested upstream: https://gitlab.gnome.org/GNOME/glib/-/merge_requests/2866#note_1544420

**Comment by SuperSandro2000** [comment]
> I'm not sure what you mean, but my goal here is just to make it possible to run the tests at all, not to make them all pass.

They don't need to pass all but we should disable the 3 tests in total that fail on other platforms.

**Review by jtojnar — APPROVED** [review]
Sounds good to me with explanation from https://gitlab.gnome.org/GNOME/glib/-/merge_requests/2866#note_1544862

**Comment by alyssais** [comment]
> They don't need to pass all but we should disable the 3 tests in total that fail on other platforms.

We shouldn't be disabling tests just because they fail.  We should be investigating them, and then either disabling them if they're Nix-specific, or reporting them upstream if they are not.  As these failures are for architectures I don't use, I can't do that, but that's fine, because these tests are not run by default at the moment anyway.  (For test failures that happen on hardware I do have access to, I [have done the upstream work](https://gitlab.gnome.org/GNOME/glib/-/merge_requests/2866).)

**Pull request merged**